engine/src/actions.ts

```diff
@@ -1,40 +1,113 @@
+import { AvailableFreeAction, freeActionData } from "./available-command";
 import { BoardAction } from "./enums";
+import Player from "./player";
+
 // QIC to extend range is already included in the distance calculation
 
-export const freeActions = [
-  { cost: "4pw", income: "1q" },
-  { cost: "3pw", income: "1o" },
-  { cost: "1q", income: "1o" },
-  { cost: "4pw", income: "1k" },
-  { cost: "1pw", income: "1c" },
-  { cost: "1k", income: "1c" },
-  { cost: "1o", income: "1c" },
-  { cost: "1o", income: "1t" },
-];
-
-export const freeActionsHadschHallas = [
-  { cost: "4c", income: "1q" },
-  { cost: "3c", income: "1o" },
-  { cost: "4c", income: "1k" },
-];
-
-export const freeActionsTerrans = [
-  { cost: "4tg", income: "1q" },
-  { cost: "3tg", income: "1o" },
-  { cost: "4tg", income: "1k" },
-  { cost: "1tg", income: "1c" },
-];
-
-export const freeActionsItars = [{ cost: "4tg", income: "tech" }];
-
-export const freeActionsNevlas = [{ cost: "1t-a3", income: "1k" }];
-
-export const freeActionsNevlasPI = [
-  { cost: "4pw", income: "1o,1c" },
-  { cost: "6pw", income: "2o" },
-];
-
-export const freeActionsBaltaks = [{ cost: "1gf", income: "1q" }];
+export enum FreeAction {
+  PowerToQic,
+  PowerToKnowledge,
+  PowerToOre,
+  PowerToCredit,
+  QicToOre,
+  OreToToken,
+  KnowledgeToCredit,
+  OreToCredit,
+
+  //HadschHallas
+  CreditToQic,
+  CreditToOre,
+  CreditToKnowledge,
+
+  //Terrans
+  GaiaTokenToQic,
+  GaiaTokenToKnowledge,
+  GaiaTokenToOre,
+  GaiaTokenToCredit,
+
+  //Itars
+  GaiaTokenToTech,
+
+  //Nevlas
+  PowerToGaiaForKnowledge,
+  PowerToOreAndCredit,
+  PowerToDoubleCredit,
+  PowerTo2Ore,
+
+  //Baltaks
+  GaiaFormerToQic,
+
+  //Taklons
+  PowerToTripleCredit,
+}
```

Comment:
Those enums are sent to the viewer in available commands. So they should have a string representation rather than the default number, or an engine upgrade could break compatibility on running games, if a free action was inserted in the middle.

Reply:
OK, I'll just not send that enum  in the command at all - just get it back based on cost/income - which is unique.